src/util/CircularBuffer.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include <vector>
+
+namespace stellar
+{
+
+using std::vector;
+using std::size_t;
+
+template <typename T> class CircularBuffer
+{
+  public:
+    explicit CircularBuffer(const size_t size) : buffer(size), count(0), start(0)
+    {
+    }
+
+    void
+    push(T elem)
+    {
+        size_t ix = postIncrementLastIndex();
+        buffer[ix] = elem;
+    }
+
+    T&
+    pop()
+    {
+        if (empty())
+        {
+            return buffer[0];
+        }
+        size_t start = postIncrementStartIndex();
+        --count;
+        return buffer[start];
+    }
+
+    bool
+    empty() const
+    {
+        return 0 == count;
+    }
+
+  private:
+    vector<T> buffer;
+    size_t count;
+    size_t start;
+
+    size_t
+    postIncrementLastIndex()
```

Comment:
"empty = (begin == end)" means full and empty states are indistinguishable. One possible solution is to add a new member describing current size.